python/rubin/rag/extra_scripts/scrape_github.py

```diff
@@ -117,38 +82,414 @@ def repos_in_org(org_name: str = "lsst-dm") -> list[str]:
+        "--single-branch",
+        "--depth",
+        "1",
+        repository_url,
+    ]
+    subprocess.run(
+        command,
+        capture_output=True,
+        text=True,
+        check=False,
+    )  # Capture output as text
+
+
+def delete_clone(repo_basename: str) -> None:
+    """Delete a GitHub repo clone from local disk.
+       Be extremely careful with this function as it
+       spawns an rm -rf command.
+
+    Parameters
+    ----------
+    repo_basename: str
+        repo base name e.g., daf_butler (does not include the org name)
+    """
+    clone_path = Path(repo_basename)
+    if clone_path.exists() and (Path.cwd() in clone_path.resolve().parents):
+        command = ["rm", "-rf", repo_basename]
+        subprocess.run(command, check=False)
+
+
+def select_doc_loader(
+    fname: str,
+) -> BSHTMLLoader | NotebookLoader | TextLoader:
+    """Select which LangChain document loader to use for a file.
+
+    Parameters
+    ----------
+    fname : str
+        path of file name that will be loaded into a LangChain doc.
+    """
+    suffix = Path(fname).suffix
+
+    if suffix == ".ipynb":
+        return NotebookLoader(fname, remove_newline=True)
+    elif suffix == ".html":
+        return BSHTMLLoader(fname)
+    else:
+        return TextLoader(fname, encoding="utf-8")
+
+
+def scrape_repo(
+    repo_name: str = "lsst/daf_butler", max_mb: int = 1024
+) -> None:
+    """
+    Scrape all non-hidden files in a locally cloned repo and batch
+    them into pickle files.
+
+    Parameters
+    ----------
+    repo_name : str
+        GitHub repository in the format 'org/repo'.
+    max_mb : int
+        Maximum size of each pickle file in megabytes.
+    """
+    # At start of scrape_repo
+
+    repo_org, repo_basename = repo_name.split("/", 1)
+    output_dir = Path(f"batched_github_output/{repo_org}/{repo_basename}")
+    if any(output_dir.glob(f"{repo_basename}_*.pkl")):
+        _log.info(f"Skipping {repo_name}, already has pickle files.")
+        return
+
+    # Extract repo organization and name
+    if "/" not in repo_name:
+        raise ValueError("Repository name should be in format 'org/repo'")
+
+    # Clone the repository
+    clone_repo(repo_name=repo_name)
+
+    # Get list of files
+    flist = clean_file_list(directory=repo_basename)
+
+    # Create output directory
+    output_dir.mkdir(parents=True, exist_ok=True)
+
+    # Process files
+    docs = []
+    current_batch: list = []
+    current_batch_size = 0
+    batch_number = 1
+    batch_size_limit = max_mb * 1024 * 1024  # Convert MB to bytes
+
+    for i, f in enumerate(flist):
+        _log.debug(f"working on file {i}, {f}")
+        loader = select_doc_loader(f)
+
+        try:
+            results = loader.load()
+            doc = results[0]
+            doc.metadata["source_key"] = "github"
+            doc.metadata["repo_basename"] = repo_basename
+            doc.metadata["org_name"] = repo_org
+            doc.metadata["repo"] = repo_name
+
```

Comment:
I can't recall if GitHub automatically does this, but we should have a "source" metadata specified for every doc. I think it would be as simple as:

```suggestion
doc.metadata["source"] = f"https://github.com/{repo_name}"
```

Reply:
Going to table this one for now as I believe Connor is out of office. Can always revisit with a future ticket/branch.